### PR TITLE
replace z3py with smtlib

### PR DESCRIPTION
**Pull request by constracti**
Replace Z3 Python API with a class which can potentially communicate with every solver supporting the SMTLIB language.

There is still some work to be done in order to reach the functionality of the master branch. Specifically, `complex_spec:f/1` test won't pass. This issue is planned to be solved when the tool will recognize custom datatypes. Moreover, smt atoms `true` and `false` are not strictly equivalent with their corresponding boolean values. Several improvements can take place (BIFs like `bit_sz` and `byte_sz`), as recursive functions are present in smt solver. However, the review process may begin.

### CODE REVIEW

**Comment by kostis** [comment]
One easy thing that can be done while I find time to review this is that it can rebased on the current `master` which has some more commits and, more importantly, tests

**Comment by kostis** [comment]
One thing that is most likely wrong in this branch is that I see duplicate results in some tests.  Take a look at e.g. test#32 and test#53 (where the issue is more pronounced).  Fundamentally, we should not be outputting the same error witness multiple times.  Somebody should look into this,

**Comment by constracti** [comment]
Isn't it possible for different constraint sets to result in the same model as a solution? Different constraint sets means different execution paths. If a "bug" is revealed when following different execution paths, we could say it is more "bugs" than one.
On the other hand, maybe the problem arises because it is (for the time being) / was impossible to express an Erlang constraint as a set of assertions in SMTLIB. For example, in test#54 (`bitstr:byte_sz/1`) cuter is trying to find a bitstring having length more than 8 by matching its length. However, from the most recent version of z3 (where recursive functions are supported), there is a more appropriate way of encoding this type of constraints. Improvements like this are planned to be implemented.

**Comment by constracti** [comment]
Issue is fixed. As a recent commit mentions, now `cuter_smt.py` implements `bitmatch_*` methods exactly as `cuter_z3.py`. It seems that previously cuter was lead (by the solver output) to explore the same execution path multiple times: various method definitions weren't clear or consistent, therefore some assumptions made caused the specific problem.

File: src/cuter_pp.erl [hunk 1]
```diff
@@ -725,7 +725,11 @@ pp_arguments(Args) ->
 pp_argument(X) ->
   case cuter_lib:is_lambda(X) of
     true  -> pp_lambda(X);
-    false -> io_lib:format("~p", [X])
```
Comment (kostis) [inline src/cuter_pp.erl]: This code can be simplified to:
```erlang
false -> io_lib:format("~p", [cuter_lib:compile_lambda(X)])
```
which will avoid duplication of similar-looking code and will also allow to remove the export of the `default_any_value/0` function from `cuter_lib`.

**Pull request merged**